pycuda/gpuarray.py

```diff
@@ -1767,6 +1767,104 @@ def make_func_for_chunk_size(chunk_size):
+    # {{{ find properties of result array
+    """
+    Join a sequence of arrays along an existing axis.
+    :arg arrays: A sequnce of :class:`GPUArray`.
+    :arg axis: Index of the dimension of the new axis in the result array.
+        Can be -1, for the new axis to be last dimension.
+    :returns: :class:`GPUArray`
+    """
+    shape = None
+    def shape_except_axis(ary: GPUArray):
+        return ary.shape[:axis] + ary.shape[axis+1:]
+
+    for i_ary, ary in enumerate(arrays):
+        allocator = allocator or ary.allocator
+
+        if shape is None:
+            # first array
+            shape = list(ary.shape)
+
+        else:
+            if len(ary.shape) != len(shape):
+                raise ValueError("%d'th array has different number of axes "
+                        "(shold have %d, has %d)"
+                        % (i_ary, len(ary.shape), len(shape)))
+
+            if (ary.ndim != arrays[0].ndim
+                    or shape_except_axis(ary) != shape_except_axis(arrays[0])):
+                raise ValueError("%d'th array has residual not matching "
+                        "other arrays" % i_ary)
+
+            shape[axis] += ary.shape[axis]
+
+    # }}}
+    shape = tuple(shape)
+    dtype = np.find_common_type([ary.dtype for ary in arrays], [])
+    result = empty(shape, dtype, allocator=allocator)
+
+    full_slice = (slice(None),) * len(shape)
+
+    base_idx = 0
+    for ary in arrays:
+        my_len = ary.shape[axis]
+        result[full_slice[:axis] + (slice(base_idx, base_idx+my_len),) + full_slice[axis+1:]] = ary
+        base_idx += my_len
+
+    return result
+
+
+def stack(arrays, axis=0, allocator=None):
+    """
+    implementation is borrowed from pyopencl.array.stack()
+
+    Join a sequence of arrays along a new axis.
+    :arg arrays: A sequnce of :class:`GPUArray`.
+    :arg axis: Index of the dimension of the new axis in the result array.
+        Can be -1, for the new axis to be last dimension.
+    :returns: :class:`GPUArray`
+    
+    """
+    allocator = allocator or arrays[0].allocator  
+
+    if not arrays:
+        raise ValueError("need at least one array to stack")
+
+    input_shape = arrays[0].shape
+    input_ndim = arrays[0].ndim
+    axis = input_ndim if axis == -1 else axis
+
+    if not all(ary.shape == input_shape for ary in arrays[1:]):
+        raise ValueError("arrays must have the same shape")
+
+    if not (0 <= axis <= input_ndim):
+        raise ValueError("invalid axis")
+
+    if (axis == 0 and not all(ary.flags.c_contiguous
+                                       for ary in arrays)):
+        # pycuda.GPUArray.__setitem__ does not support non-contiguous assignments
+        raise NotImplementedError
+
+    if (axis == input_ndim and not all(ary.flags.f_contiguous
+                                               for ary in arrays)):
+        # pycuda.Array.__setitem__ does not support non-contiguous assignments
+        raise NotImplementedError
```

Comment:
Are you sure pycuda arrays do not support non-contiguous array assignment?
```python
In [9]: import numpy as np

In [10]: import pycuda.gpuarray as gpuarray

In [11]: import pycuda.autoinit

In [12]: a_gpu = gpuarray.zeros((10, ), dtype=np.float32)

In [13]: a_gpu[::2] = 42*np.ones(5, dtype=np.float32)

In [14]: a_gpu
Out[14]: array([42.,  0., 42.,  0., 42.,  0., 42.,  0., 42.,  0.], dtype=float32)
```

Reply:
Ohh got it. In that case let me get rid of those 2 checks then. Thanks for the quick demo